File: Service-User/src/main/java/org/example/serviceuser/util/PasswordValidator.java
/**
 * File Name: PasswordValidator.java
 * Description: Todo
 * Author: holic512
 * Created Date: 2025-05-10
 * Version: 1.0
 * Usage:
 * Todo
 */
package org.example.serviceuser.util;

// PasswordValidator.java

import java.util.*;
import java.util.regex.Pattern;

public final class PasswordValidator {

    // 提供一个默认实例，方便快捷调用
    public static final PasswordValidator DEFAULT =
            new PasswordValidator(new Options.Builder().build());
    // 特殊字符正则表达式
    private static final Pattern SPECIAL_CHARS_PATTERN = Pattern.compile("[!@#$%^&*()\\-_=+\\[\\]{};:'\",.<>/?\\\\|]");
    private final Options cfg;

    public PasswordValidator(Options options) {
        this.cfg = options;
    }

    /**
     * 核心校验方法：单次遍历字符串，收集所有信息
     */
    public ValidationResult validate(String pwd) {
        List<String> errors = new ArrayList<>(4);

        // 预处理输入，处理空字符串和null
        if (pwd == null) {
            errors.add("密码不能为空");
            return new ValidationResult(false, errors);
        }

        // 去除前后空格
        pwd = pwd.trim();
        int len = pwd.length();

        // 空密码检查
        if (len == 0) {
            errors.add("密码不能为空");
            return new ValidationResult(false, errors);
        }

        // 1. 长度校验
        if (len < cfg.minLength || len > cfg.maxLength) {
            errors.add(
                    String.format("长度必须在 %d 到 %d 个字符之间", cfg.minLength, cfg.maxLength)
            );
        }

        // 遍历标志
        boolean hasUpper = false, hasLower = false, hasDigit = false;
        boolean hasSpecial = false, hasWhitespace = false;
        boolean repeatErrorReported = false;
        char last = '\0';
        int consecCount = 1;

        for (int i = 0; i < len; i++) {
            char c = pwd.charAt(i);
            if (Character.isUpperCase(c)) hasUpper = true;
            else if (Character.isLowerCase(c)) hasLower = true;
            else if (Character.isDigit(c)) hasDigit = true;
            else if (Character.isWhitespace(c)) {
                hasWhitespace = true;
            } else if (SPECIAL_CHARS_PATTERN.matcher(String.valueOf(c)).matches()) {
                hasSpecial = true;
            }

            // 检查连续重复 - 修正逻辑
            if (i > 0 && c == last) {
                consecCount++;
            } else {
                consecCount = 1;
            }

            // 报告连续重复错误（如果未报告过）
            if (!repeatErrorReported && consecCount > cfg.maxConsecutiveRepeat) {
                errors.add(
                        String.format("不能有超过 %d 个连续相同字符", cfg.maxConsecutiveRepeat)
                );
                repeatErrorReported = true;
            }

            last = c;
        }

        // 2. 空白字符
        if (cfg.noWhitespace && hasWhitespace) {
            errors.add("不能包含空白字符");
        }
        // 3. 大小写、数字、特殊字符
        if (cfg.requireUppercase && !hasUpper) {
            errors.add("必须包含至少一个大写字母 (A-Z)");
        }
        if (cfg.requireLowercase && !hasLower) {
            errors.add("必须包含至少一个小写字母 (a-z)");
        }
        if (cfg.requireDigit && !hasDigit) {
            errors.add("必须包含至少一个数字 (0-9)");
        }
        if (cfg.requireSpecial && !hasSpecial) {
            errors.add("必须包含至少一个特殊字符（如 !@#$%^&*() 等）");
        }

        // 4. 弱密码/违禁密码检测
        String pwdLower = pwd.toLowerCase();

        // 根据配置选择全匹配或子串匹配
        if (cfg.useBannedSubstringMatch) {
            // 子串匹配 - 检查密码是否包含任何违禁词
            for (String banned : cfg.bannedSet) {
                if (pwdLower.contains(banned)) {
                    errors.add("密码包含常见弱密码或违禁词");
                    break;
                }
            }
        } else {
            // 全匹配 - 检查密码是否与任何违禁词完全相同
            if (cfg.bannedSet.contains(pwdLower)) {
                errors.add("禁止使用弱密码或违禁密码");
            }
        }

        boolean valid = errors.isEmpty();
        return new ValidationResult(valid, errors);
    }

    // 校验结果
    public record ValidationResult(boolean valid, List<String> errors) {
    }

    // 可配置项
    public static class Options {
        public final int minLength;
        public final int maxLength;
        public final boolean requireUppercase;
        public final boolean requireLowercase;
        public final boolean requireDigit;
        public final boolean requireSpecial;
        public final boolean noWhitespace;
        public final int maxConsecutiveRepeat;
        public final Set<String> bannedSet;
        public final boolean useBannedSubstringMatch;

        private Options(Builder b) {
            this.minLength = b.minLength;
            this.maxLength = b.maxLength;
            this.requireUppercase = b.requireUppercase;
            this.requireLowercase = b.requireLowercase;
            this.requireDigit = b.requireDigit;
            this.requireSpecial = b.requireSpecial;
            this.noWhitespace = b.noWhitespace;
            this.maxConsecutiveRepeat = b.maxConsecutiveRepeat;
            this.useBannedSubstringMatch = b.useBannedSubstringMatch;
            // 统一小写并构造 HashSet，便于快速查找
            Set<String> tmp = new HashSet<>();
            for (String p : b.bannedList) {
                if (p != null && !p.isEmpty()) {
                    tmp.add(p.toLowerCase());
                }
            }
            this.bannedSet = Collections.unmodifiableSet(tmp);
        }

        public static class Builder {
            private int minLength = 8;
            private int maxLength = 32;
            private boolean requireUppercase = true;
            private boolean requireLowercase = true;
            private boolean requireDigit = true;
            private boolean requireSpecial = true;
            private boolean noWhitespace = true;
            private int maxConsecutiveRepeat = 2;
            private boolean useBannedSubstringMatch = true;
            private List<String> bannedList = List.of(
                    "password", "123456", "12345678", "123456789", "1234567890",
                    "qwerty", "qwerty123", "abc123", "abc12345", "111111", "11111111",
                    "123123", "121212", "password1", "passw0rd", "letmein", "welcome",
                    "dragon", "sunshine", "master", "football", "baseball", "monkey",
                    "shadow", "superman", "michael", "mustang", "password123",
                    "pokemon", "iloveyou", "696969", "1234", "12345", "1234abcd",
                    "000000", "qazwsx", "123qwe", "123abc", "1q2w3e4r", "zaq1zaq1",
                    "admin", "administrator", "root", "sysadmin", "support", "staff",
                    "pass123", "login", "hello", "freedom", "whatever", "trustno1",
                    "flower", "hottie", "lovely", "zaq12wsx", "baseball1"
            );

            public Builder minLength(int len) {
                this.minLength = len;
                return this;
            }

            public Builder maxLength(int len) {
                this.maxLength = len;
                return this;
            }

            public Builder requireUppercase(boolean f) {
                this.requireUppercase = f;
                return this;
            }

            public Builder requireLowercase(boolean f) {
                this.requireLowercase = f;
                return this;
            }

            public Builder requireDigit(boolean f) {
                this.requireDigit = f;
                return this;
            }

            public Builder requireSpecial(boolean f) {
                this.requireSpecial = f;
                return this;
            }

            public Builder noWhitespace(boolean f) {
                this.noWhitespace = f;
                return this;
            }

            public Builder maxConsecutiveRepeat(int n) {
                this.maxConsecutiveRepeat = n;
                return this;
            }

            public Builder bannedList(List<String> list) {
                this.bannedList = list;
                return this;
            }

            public Builder useBannedSubstringMatch(boolean f) {
                this.useBannedSubstringMatch = f;
                return this;
            }

            public Options build() {
                return new Options(this);
            }
        }
    }
}

File: frontend-user-web/src/utils/passwordValidator.ts
// passwordValidator.ts

/**
 * 校验结果接口
 */
export interface ValidationResult {
    valid: boolean;
    errors: string[];
}

/**
 * 默认的常见弱密码或违禁密码列表（可按需增删）
 */
const defaultBannedPasswords = [
    // 纯数字常见弱口令
    '1234', '12345', '123456', '1234567', '12345678', '123456789', '1234567890',
    '000000', '111111', '11111111', '121212',
    '696969', '123123', '123123123',

    // 常见字母组合
    'password', 'password1', 'passw0rd',
    'qwerty', 'qwerty123', 'qazwsx', 'zaq12wsx', 'zaq1zaq1', '123qwe',

    // 单词/短语
    'abc123', 'abc12345', 'letmein', 'welcome', 'dragon', 'sunshine',
    'master', 'superman', 'pokemon', 'football', 'baseball', 'monkey',
    'shadow', 'flower', 'hottie', 'lovely', 'freedom', 'whatever',

    // 个性化弱口令
    'admin', 'administrator', 'root', 'sysadmin', 'support', 'staff',
    'login', 'user', 'pass123', 'trustno1', 'iloveyou', '1234abcd',
    '123abc', '1q2w3e4r', 'master123', 'football1', 'baseball1'
];

// 特殊字符正则表达式
const SPECIAL_CHARS_PATTERN = /[!@#$%^&*()_\-=+[\]{};:'",.<>/?\\|]/;

/**
 * 密码校验选项
 */
export interface PasswordOptions {
    minLength?: number;           // 最小长度
    maxLength?: number;           // 最大长度
    requireUppercase?: boolean;   // 是否要求大写字母
    requireLowercase?: boolean;   // 是否要求小写字母
    requireDigit?: boolean;       // 是否要求数字
    requireSpecialChar?: boolean; // 是否要求特殊字符
    noWhitespace?: boolean;       // 是否禁止空白字符
    maxConsecutiveRepeat?: number;// 最大连续重复字符数，超过视为失败
    bannedList?: string[];        // 自定义的弱密码或违禁密码列表
    useBannedSubstringMatch?: boolean; // 是否启用违禁词子串匹配
}

/**
 * 默认密码校验规则
 */
const defaultOptions: Required<PasswordOptions> = {
    minLength: 8,
    maxLength: 32,
    requireUppercase: true,
    requireLowercase: true,
    requireDigit: true,
    requireSpecialChar: true,
    noWhitespace: true,
    maxConsecutiveRepeat: 2,
    bannedList: defaultBannedPasswords,
    useBannedSubstringMatch: true, // 默认启用子串匹配
};

/**
 * 校验密码
 * @param password 要校验的密码
 * @param options 可选校验规则配置
 * @returns {ValidationResult} 校验结果
 */
export function validatePassword(
    password: string | null | undefined,
    options: PasswordOptions = {}
): ValidationResult {
    const cfg = { ...defaultOptions, ...options };
    const errors: string[] = [];

    // 预处理输入
    if (password === null || password === undefined) {
        return {
            valid: false,
            errors: ['密码不能为空']
        };
    }
    
    // 去除前后空格
    password = password.trim();
    
    // 空密码检查
    if (password.length === 0) {
        return {
            valid: false,
            errors: ['密码不能为空']
        };
    }

    // 1. 长度校验
    if (password.length < cfg.minLength || password.length > cfg.maxLength) {
        errors.push(`长度必须在 ${cfg.minLength} 到 ${cfg.maxLength} 个字符之间`);
    }

    // 2. 大小写字母、数字、特殊字符校验
    if (cfg.requireUppercase && !/[A-Z]/.test(password)) {
        errors.push('必须包含至少一个大写字母 (A-Z)');
    }
    if (cfg.requireLowercase && !/[a-z]/.test(password)) {
        errors.push('必须包含至少一个小写字母 (a-z)');
    }
    if (cfg.requireDigit && !/[0-9]/.test(password)) {
        errors.push('必须包含至少一个数字 (0-9)');
    }
    if (cfg.requireSpecialChar && !SPECIAL_CHARS_PATTERN.test(password)) {
        errors.push('必须包含至少一个特殊字符（如 !@#$%^&*() 等）');
    }

    // 3. 禁止空白字符
    if (cfg.noWhitespace && /\s/.test(password)) {
        errors.push('不能包含空白字符');
    }

    // 4. 禁止连续重复字符过多
    if (cfg.maxConsecutiveRepeat > 0) {
        // 检查连续重复字符
        let repeatCount = 1;
        let lastChar = '';
        let hasRepeatError = false;
        
        for (let i = 0; i < password.length; i++) {
            const char = password[i];
            if (i > 0 && char === lastChar) {
                repeatCount++;
                if (!hasRepeatError && repeatCount > cfg.maxConsecutiveRepeat) {
                    errors.push(`不能有超过 ${cfg.maxConsecutiveRepeat} 个连续相同字符`);
                    hasRepeatError = true;
                }
            } else {
                repeatCount = 1;
            }
            lastChar = char;
        }
    }

    // 5. 弱密码/违禁密码检查
    const lower = password.toLowerCase();
    
    if (cfg.useBannedSubstringMatch) {
        // 子串匹配 - 检查密码是否包含任何违禁词
        for (const banned of cfg.bannedList) {
            if (banned && lower.includes(banned.toLowerCase())) {
                errors.push('密码包含常见弱密码或违禁词');
                break;
            }
        }
    } else {
        // 全匹配 - 检查密码是否与任何违禁词完全相同
        for (const banned of cfg.bannedList) {
            if (banned && lower === banned.toLowerCase()) {
                errors.push(`禁止使用弱密码 "${banned}"`);
                break;
            }
        }
    }

    return {
        valid: errors.length === 0,
        errors,
    };
}
